Normative: Define coalescing of syntactic elements

littledan opened the pull request:
When decorating a getter/setter pair, the syntax is to put all
decorators before one of the two, and then these decorators will
apply to the coalesced pair together. This patch defines how that
coalescing works, with the following semantic details:
- The definition of coalescing matches the way class definitions
  execute in ES2015, through repeated application of
  DefineOwnProperty. For example, if an ordinary method declaration
  follows a getter declaration, then the getter is silently replaced
  by the method.
- Duplicate same-named field declarations, or a field declaration
  with the same name as a method declaration, are allowed and
  don't trigger coalescing.
- If both a getter and a setter, or any other pair of same-named
  methods, are both decorated, then a ReferenceError is thrown
  at runtime. We may want to create an early error sometimes
  in a follow-on patch (#60), but the runtime error exists at
  least for the computed property name case.

Closes #52

littledan committed:
Normative: Define coalescing of syntactic elements

When decorating a getter/setter pair, the syntax is to put all
decorators before one of the two, and then these decorators will
apply to the coalesced pair together. This patch defines how that
coalescing works, with the following semantic details:
- The definition of coalescing matches the way class definitions
  execute in ES2015, through repeated application of
  DefineOwnProperty. For example, if an ordinary method declaration
  follows a getter declaration, then the getter is silently replaced
  by the method.
- Duplicate same-named field declarations, or a field declaration
  with the same name as a method declaration, are allowed and
  don't trigger coalescing.
- If both a getter and a setter, or any other pair of same-named
  methods, are both decorated, then a ReferenceError is thrown
  at runtime. We may want to create an early error sometimes
  in a follow-on patch (#60), but the runtime error exists at
  least for the computed property name case.

Closes #52

ljharb commented on spec.html:
```diff
@@ -125,6 +125,38 @@ <h1>Updated Productions</h1>
+            1. Set _other_.[[Decorators]] to _element_.[[Decorators]].
+          1. If _element_.[[Key]] is a Private Name,
+            1. Assert: _element_.[[Kind]] is `"method"`.
+            1. If _element_.[[Descriptor]] has a [[Get]] field,
```
Could we use abstract operations here, like IsAccessorDescriptor/isMutatorDescriptor or something similar? I feel like this algorithm here would be more readable if all the shape assertions and branching were extracted out.

littledan replied:
Do you have a particular other wording that you'd like to propose?

ljharb replied:
rereading it, there's not much else to be done with existing abstract ops :-/

we could make new tiny ones but i doubt it would aid much of the readability here.

Maybe we could make a new abstract op like `CoalesceAccessorDescriptor(_element_, _other_)`, which could contain the assertions and the branching, and it'd simplify the algorithm in `CoalesceClassElements`?

littledan replied:
I'm not sure if increasing the indirection here will make this algorithm more readable. You're talking about factoring out just those two lines for Get and Set?

ljharb replied:
My thinking is that it would at a minimum collapse lines 288-294 to "assert isDataDescriptor + CoalesceAccessorDescriptor()", and lines lines 300-302 to the same.

littledan replied:
Oh, I see, somehow I missed the repetition between the private and public cases. Sure, I'll make this change.

littledan committed:
Editorial: Simplify definition of coalescing

- Factor out coalescing of getter/setter pairs
- Reduce the separation of code paths between private names
  and ordinary properties (the logic is actually the same)

littledan commented:
OK, I've updated the specification to factor out coalescing of getter/setter pairs along the lines @ljharb suggested.

ljharb commented on spec.html:
```diff
@@ -273,6 +273,40 @@ <h1>Runtime Semantics: BindingClassDeclarationEvaluation</h1>
+    <emu-alg>
+      1. Assert: IsAccessorDescriptor(_other_.[[Descriptor]]) and IsAccessorDescriptor(_element_.[[Descriptor]]) are both *true*,
+      1. If _element_.[[Descriptor]] has a [[Get]] field,
+        1. Set _other_.[[Descriptor]].[[Get]] to _element_.[[Descriptor]].[[Get]].
```
Is it worth asserting that other does not have a Get field?

littledan replied:
That assertion might not be valid for public methods (where we don't have the benefit of early errors for misuse in conjunction with computed property names).

ljharb commented on spec.html:
```diff
@@ -273,6 +273,40 @@ <h1>Runtime Semantics: BindingClassDeclarationEvaluation</h1>
+        1. Set _other_.[[Descriptor]].[[Get]] to _element_.[[Descriptor]].[[Get]].
+      1. Otherwise,
+        1. Assert: _element_.[[Descriptor]] has a [[Set]] field.
+        1. Set _other_.[[Descriptor]].[[Set]] to _element_.[[Descriptor]].[[Set]].
```
And here, to assert that other does not have a Set?

ljharb commented on spec.html:
```diff
@@ -273,6 +273,40 @@ <h1>Runtime Semantics: BindingClassDeclarationEvaluation</h1>
+            1. Set _other_.[[Decorators]] to _element_.[[Decorators]].
+          1. If IsDataDescriptor(_element_.[[Descriptor]]) is *true* or IsDataDescriptor(_other_.[[Descriptor]]) is *true*, then
+            1. Assert: _element_.[[Key]] is not a Private Name.
+            1. Assert: _element_.[[Descriptor]].[[Configurable]] is *true*, and _other_.[[Descriptor]].[[Configurable]] is *true*.
```
Is there no possibility that a decorator could return a nonconfigurable public data descriptor with kind “method” here?

littledan replied:
No, since this is only run on syntactic constructs before decorators.

ljharb reviewed (commented):
Thanks, this seems much easier for me to understand :-)

wycats reviewed (approved):
This looks good to me. We might want to loosen the "decorate twice" restriction if we find that people are repeatedly butting up against patterns where it makes sense for some reason, but I agree that this is best.
